[webview_flutter]Support UserAgent setting for webview_flutter

YukiOya opened the pull request:
It possible to get and set custom user agent to webview.

ened commented on packages/webview_flutter/ios/Classes/FlutterWebView.m:
```diff
@@ -88,11 +90,30 @@ - (void)onLoadUrl:(FlutterMethodCall*)call result:(FlutterResult)result {
+               if (error) {
+                 result([FlutterError
+                     errorWithCode:@"getUserAgent_failed"
+                           message:@"Failed gettting UserAgent"
```
typo

ened commented on packages/webview_flutter/ios/Classes/FlutterWebView.m:
```diff
@@ -113,6 +134,16 @@ - (void)updateJsMode:(NSNumber*)mode {
+  if (@available(iOS 9.0, *)) {
+    [_webView setCustomUserAgent:userAgent];
+  } else if (userAgent) {
+    [[NSUserDefaults standardUserDefaults] registerDefaults:@{@"UserAgent" : userAgent}];
```
Why is this required? Where is the value read again?

amirh replied:
Looks like `WkWebView` reads that value on construction, I did not find official documentation for this other than SO, but I did verify that it works.

ened replied:
In that case.. I’d recommend a short source code comment, perhaps a link to the relevant SO answer.

amirh commented on packages/webview_flutter/ios/Classes/FlutterWebView.m:
```diff
@@ -113,6 +134,16 @@ - (void)updateJsMode:(NSNumber*)mode {
+  } else if (userAgent) {
+    [[NSUserDefaults standardUserDefaults] registerDefaults:@{@"UserAgent" : userAgent}];
+    WKWebViewConfiguration* configuration = _webView.configuration;
+    _webView = [[WKWebView alloc] initWithFrame:_webView.frame configuration:configuration];
```
We can't do this as it won't change the webview reference that the platform views controller is holding.
Even if we could update the engine's reference, doing this will lose any state the webview is currently in.

I think a better approach for iOS < 9 would be to do the user defaults trick prior to the  construction of the webview in `FLTWebViewController`'s `initWithFrame` (ohh I just noticed we have a typo there).

And just not support changing the userAgent for iOS < 9 (as WkWebView doesn't support it).

Note that we should make sure to restore the UserAgent value after constructing the webview.

YukiOya commented:
@amirh Thanks for the review.
 I fixed the issues, and added new commit to fix the accessor name like [currentUrl](https://github.com/flutter/plugins/pull/992).

msdk27 commented:
Hey, thanks for the missing part in the webview plugin... my question is, how to use the own userAgent?! Is there a documentation?

YukiOya commented:
@msdk27 Hi. Thank you for your question.
Now, the custom userAgent apply to webview only if it initialize with constructor argument like this.

```dart
WebView(userAgent: "custom useragent");
```

Perhaps, the interface will change by maintainer's reviews.

msdk27 commented:
> @msdk27 Hi. Thank you for your question.
> Now, the custom userAgent apply to webview only if it initialize with constructor argument like this.
> 
> ```dart
> WebView(userAgent: "custom useragent");
> ```
> 
> Perhaps, the interface will change by maintainer's reviews.

Hi thank you for response. Do you have an example? it does not work for me.

amirh commented:
Sorry for the delay getting to review this, we were trying to minimize adding new features before we had e2e tests.

The code has changed a bit since this PR was sent, this PR needs to be rebased, and add e2e tests should be added(see https://github.com/flutter/plugins/blob/master/packages/webview_flutter/example/test_driver/webview.dart).

YukiOya commented:
Hi, @amirh 
Thank you for reply.
This PR is rebased, and added e2e tests.

amirh commented on packages/webview_flutter/lib/webview_flutter.dart:
```diff
@@ -205,6 +206,9 @@ class WebView extends StatefulWidget {
   /// [WebViewController.evaluateJavascript] can assume this.
   final PageFinishedCallback onPageFinished;
 
+  /// The custom UserAgent.
```
We should have more information in this Dartdoc, some question I think I might be having if I were looking here:

1. What would be the user agent when this is null?
2. What happens when the widget is rebuild with a new userAgent? is it sending a new request with the new user agent?
3. What happens when `goBack` is called after a user agent was changed? is the new or previous user agent used?

In general I recommend reading the [documentation section in the Flutter style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#documentation-dartdocs-javadocs-etc) which we should follow here.

amirh commented on packages/webview_flutter/lib/webview_flutter.dart:
```diff
@@ -490,6 +501,11 @@ class WebViewController {
   }
 
+  /// Accessor the UserAgent.
+  Future<String> userAgent() async {
```
nit: getUserAgent

amirh commented on packages/webview_flutter/lib/webview_flutter.dart:
```diff
@@ -490,6 +501,11 @@ class WebViewController {
     return _channel.invokeMethod("reload");
   }
 
+  /// Accessor the UserAgent.
```
Please follow the [documentation section in the Flutter style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#documentation-dartdocs-javadocs-etc), I'd suggest:

```
/// Returns the User-Agent value that will be used for subsequent HTTP requests.
```

amirh commented on packages/webview_flutter/example/test_driver/webview.dart:
```diff
@@ -54,6 +54,28 @@ void main() {
+    final String userAgent = await controller.userAgent();
+    expect(userAgent, 'UA');
+  });
 }
```
Please add a test that rebuilds a WebView with a different user agent.

timrijckaert commented:
What needs to be done to get this merged?

aegis123 commented:
@YukiOya any change you can get the last remarks fixed so this can be merged?

amirh commented:
This PR needs more work before it can be merged, a rebase/merge of master is needed(a recent refactoring that was done requires moving some stuff around, and following up on the review comments is still needed).

I'm marking this PR as "needs love" and closing it, feel free to re-open or file a new PR.
Note that the priority of the feature(and of reviewing non trivial PRs) is determined by the number thumbs up on the issue(https://github.com/flutter/flutter/issues/28256 in this case). See the [review process](https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#the-review-process) section in the contributing guide.
